src/zarr.encoder.hh

```diff
@@ -0,0 +1,41 @@
+#ifndef H_ACQUIRE_STORAGE_ZARR_ENCODER_V0
+#define H_ACQUIRE_STORAGE_ZARR_ENCODER_V0
+
+#ifdef __cplusplus
+
+#include <string>
+#include <vector>
+
+#include "platform.h"
+
+#include "prelude.h"
+#include "json.hpp"
+
+namespace acquire::sink::zarr {
+
+struct BaseEncoder
+{
+  public:
+    BaseEncoder();
+    virtual ~BaseEncoder() noexcept = default;
+
+    size_t write(const uint8_t* beg, const uint8_t* end);
+    size_t flush();
+    void set_bytes_per_pixel(size_t bpp);
+    void allocate_buffer(size_t buf_size);
+
+    virtual void set_file(struct file* file_handle);
+
+  protected:
+    std::vector<uint8_t> buf_;
+    size_t cursor_;
+    size_t bytes_per_pixel_;
+    std::string path_;
+    struct file* file_; // non-owning
```

Comment:
I don't think this merit's a change, but the need for this file_ pointer here is awkward. It's a consequence of using inheritance as a composition mechanism here.

It looks like the owner of the file is always a child class. This guarantees the lifetime of the pointer except in the destructor - the parent gets destroyed last - but things look safe there at the moment.

Anyway, no action here just wanted to raise it just in case.

Reply:
The `ChunkWriter` owns the file in a `std::optional` wrapper. Either the file is open, so it [has a value](https://github.com/aliddell/acquire-driver-zarr/blob/nontrivial-chunking/src/chunk.writer.cpp#L193) and the Encoder gets a pointer to it, or the file is closed so it [doesn't have a value](https://github.com/aliddell/acquire-driver-zarr/blob/nontrivial-chunking/src/chunk.writer.cpp#L215) and the pointer is null.